jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/profiles/ProfileWatcher.kt

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials.profiles
+
+import com.intellij.util.io.isFile
+import software.amazon.awssdk.profiles.ProfileFileLocation
+import software.aws.toolkits.core.utils.debug
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.info
+import software.aws.toolkits.core.utils.warn
+import java.nio.file.FileSystems
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardWatchEventKinds
+import java.nio.file.WatchKey
+
+internal class ProfileWatcher(private val onChange: () -> Unit) {
+    private val watchService = FileSystems.getDefault().newWatchService()
+    private val watchLocations = setOf(
+        ProfileFileLocation.configurationFilePath().toAbsolutePath(),
+        ProfileFileLocation.configurationFilePath().toAbsolutePath()
+    )
+    private val watchKeys: MutableMap<WatchKey, Path> = mutableMapOf()
+
+    fun start() {
+        watchLocations.filter {
+            if (Files.exists(it)) {
+                true
+            } else {
+                LOGGER.info { "$it does not exist, won't watch" }
+                false
+            }
+        }.map {
+            val watchLocation = if (it.isFile()) it.parent else it
+            watchLocation.register(
+                watchService,
+                StandardWatchEventKinds.ENTRY_CREATE,
+                StandardWatchEventKinds.ENTRY_DELETE,
+                StandardWatchEventKinds.ENTRY_MODIFY
+            ) to watchLocation
+        }.forEach {
+            LOGGER.info { "Watching ${it.second} for file changes" }
+            watchKeys[it.first] = it.second
+        }
+
+        while (true) {
+            val key = try {
+                watchService.take()
+            } catch (e: InterruptedException) {
+                return
+            }
+
+            key.pollEvents().forEach {
+                try {
+                    val kind = it.kind()
+                    when (kind) {
+                        StandardWatchEventKinds.OVERFLOW -> {
+                            LOGGER.debug { "ProfileWatcher got an OVERFLOW" }
+                        }
+                        else -> {
+                            // Context path is relative to base registered to WatchKey
+                            val context = it.context()
+                            if (context is Path) {
+                                val fullPath = watchKeys[key]?.resolve(context)
+                                LOGGER.debug { "$fullPath was changed" }
+                                if (watchLocations.contains(fullPath)) {
+                                    onChange()
```

Comment:
Have we decided that a partial file refresh is not practical?

Reply:
We should come back and do it, so that MFA for unaffected profiles arent affected. Right now, I will just do a full refresh.